orders-accounting/src/cache.rs

```diff
@@ -0,0 +1,258 @@
+            Error::InvariantOrderDataExistForWithdrawUndo(undo.id)
+        );
+        self.data.order_data.undo_merge_delta_data_element(undo.id, undo.undo_data)?;
+
+        ensure!(
+            self.get_ask_balance(&undo.id)?.unwrap_or(Amount::ZERO) == Amount::ZERO,
+            Error::InvariantOrderAskBalanceExistForWithdrawUndo(undo.id)
+        );
+        self.data.ask_balances.add_unsigned(undo.id, undo.ask_balance)?;
+
+        ensure!(
+            self.get_give_balance(&undo.id)?.unwrap_or(Amount::ZERO) == Amount::ZERO,
+            Error::InvariantOrderGiveBalanceExistForWithdrawUndo(undo.id)
+        );
+        self.data.give_balances.add_unsigned(undo.id, undo.give_balance)?;
+
+        Ok(())
+    }
+
+    fn undo_fill_order(&mut self, undo: FillOrderUndo) -> Result<()> {
+        if let Some(undo_data) = undo.undo_data {
+            ensure!(
+                self.get_order_data(&undo.id)?.is_none(),
+                Error::InvariantOrderDataExistForWithdrawUndo(undo.id)
+            );
+            self.data.order_data.undo_merge_delta_data_element(undo.id, undo_data)?;
+        }
+
+        self.data.ask_balances.add_unsigned(undo.id, undo.ask_balance)?;
+        self.data.give_balances.add_unsigned(undo.id, undo.give_balance)?;
+
+        Ok(())
+    }
+}
+
+impl<P: OrdersAccountingView> OrdersAccountingView for OrdersAccountingCache<P> {
+    type Error = Error;
+
+    fn get_order_data(&self, id: &OrderId) -> Result<Option<OrderData>> {
+        match self.data.order_data.get_data(id) {
+            accounting::GetDataResult::Present(d) => Ok(Some(d.clone())),
+            accounting::GetDataResult::Deleted => Ok(None),
+            accounting::GetDataResult::Missing => {
+                Ok(self.parent.get_order_data(id).map_err(|_| Error::ViewFail)?)
+            }
+        }
+    }
+
+    fn get_ask_balance(&self, id: &OrderId) -> Result<Option<Amount>> {
+        let parent_supply = self.parent.get_ask_balance(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.ask_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+
+    fn get_give_balance(&self, id: &OrderId) -> Result<Option<Amount>> {
+        let parent_supply = self.parent.get_give_balance(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.give_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+}
+
+impl<P: OrdersAccountingView> OrdersAccountingOperations for OrdersAccountingCache<P> {
+    fn create_order(&mut self, id: OrderId, data: OrderData) -> Result<OrdersAccountingUndo> {
+        if self.get_order_data(&id)?.is_some() {
+            return Err(Error::OrderAlreadyExists(id));
+        }
+
+        if self.get_ask_balance(&id)?.is_some() {
+            return Err(Error::OrderAlreadyExists(id));
+        }
+
+        let ask_value = data.ask().clone();
```

Comment:
do we want to forbid orders where ask and give currencies are the same?

Reply:
Yes